_posts/2015-01-20-ops-lessons-and-instant-temporary-ipython-notebooks.md

```diff
@@ -0,0 +1,110 @@
+
+* [Olivier Grisel](https://twitter.com/ogrisel) used it in his scikit learn tutorial
+
+After all the above activity, we learned several things:
+
+* Serve static assets via nginx instead of each user container
+* Websockets and proxies gobble up ports
+* People want to share notebooks
+* Pool user containers ahead of time
+
+### Static assets should be served via nginx instead of within each user container
+
+nginx is really good at serving static resources. Let it handle the grunt work easily:
+
+```
+location ~ /(user[-/][a-zA-Z0-9]*)/static/(.*) {
+    alias /srv/ipython/IPython/html/static/$2;
+}
+```
+
+This is both for speed and to use fewer file descriptors across the system. We ended up pulling some neat tricks with our deployment setup to [mount a dummy user container's static files into an nginx container](https://github.com/jupyter/tmpnb-deploy/pull/3). However, if you don't do this, you get the neat side effect of never having caching problems when launching new versions of userland containers, and even means you could ship different ones to different users!
+
+### Websockets and proxies gobble up ports
+
+Each websocket opened by a user ends up opening several ports that will stay open. Since each notebook on a notebook server will open at least one websocket, the number of notebooks you have open directly correlates with the number of websockets.
+
+Here's some trimmed and commented `lsof -i` output when a single user is accessing an IPython 3.x notebook server through tmpnb:
+
+```
+# The fixed pieces of networking for tmpnb itself
+# Node proxy
+node     2646 *:8000 (LISTEN)
+
+# Orchestration layer of tmpnb
+python   2669 *:9999 (LISTEN)
+python   2669 *:9999 (LISTEN)
+node     2646 ip6-localhost:8001 (LISTEN)
+
+# Now for the pieces that each user will be "allocated"
+# Established connection with client
+node     2646 10.184.2.134:8000->10.223.242.4:53254 (ESTABLI)
+
+# Docker exposes port for IPython notebook server
+docker  18173 ip6-localhost:49216 (LISTEN)
+
+# Proxy connects to docker container, keeping port allocated to websocket
+node     2646 ip6-localhost:35831->ip6-localhost:49216 (ESTABLI)
+docker  18173 ip6-localhost:49216->ip6-localhost:35831 (ESTABLI)
+
+# Docker routes onward to the IPython Notebook
+docker  18173 ip6-localhost:49216->ip6-localhost:35673 (ESTABLI)
+python   2669 ip6-localhost:35673->ip6-localhost:49216 (ESTABLI)
+```
+
+This got really bad when the notebook server had 3 websockets per notebook and users were opening more than a few notebooks during a tutorial. Websockets have to stay established, by nature.
+
+### People want to share notebooks
+
+Honestly, I should know this from working on the [notebook viewer](http://nbviewer.ipython.org/). On several occasions, people have passed me links on the demo version of tmpnb, hosted at tmpnb.org.
+
+```
+https://tmpnb.org/user/FKlVK3haOQRF/notebooks/SharingEffects.ipynb
+```
```

Comment:
You might not want to actually include an example link here, since it won't work?

It might also be helpful to spell out _why_ it's a bad idea - just reiterate that individual containers are ephemeral, etc. It certainly looks like it works, after all.